lib/streamlit/commands/query_params.py

```diff
@@ -143,29 +143,102 @@ def _ensure_no_embed_params(
     return current_embed_params
 
 
+def _missing_key_error_message(key: str) -> str:
+    return f'st.query_params has no key "{key}". Did you forget to initialize it? '
+
+
 @dataclass
 class QueryParams:
     """A dict-like representation of query params.
-    The main difference is that it only stores and returns str and list[str].
+    The main difference is that it only stores and returns str and List[str].
 
     TODO(willhuang1997): Fill in these docs with examples and fix doc above. Above is just a stub for now.
     """
 
-    _query_params: Dict[str, List[Any]] = field(default_factory=dict)
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self, query_params: Dict[str, Union[List[str], str]] = {}):
+        # avoid using ._query_params as that will use __setattr__
+        # and access _query_params which is None as _query_params isn't initialized
+        self.__dict__["_query_params"] = query_params
+
+    def _keys(self) -> Set[str]:
+        return set(self._query_params.keys())
+
+    def get(self, key: str, default: Any = None) -> Any:
+        if key in self._query_params:
+            return self._getitem(key)
+        else:
+            if default != None:
+                self._setitem(key, default)
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._keys())
 
-    def __init__(self, query_params: Dict[str, List[Any]] = {}):
-        self._query_params = query_params
+    def __getitem__(self, key: str) -> Any:
+        return self._getitem(key)
 
-    def __repr__(self):
-        return util.repr_(self)
+    def _getitem(self, key: str) -> str:
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return  # type: ignore
+        try:
+            value = self._query_params[key]
+            if isinstance(value, list):
+                # This should never happen as you can't set a key to an empty array
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]
+            return value
+        except:
+            raise KeyError(_missing_key_error_message(key))
+
+    def _setitem(self, key: str, value: Any) -> None:
+        if isinstance(value, list):
+            self._query_params[key] = util.convert_to_strings_in_list(value)
+        else:
+            self._query_params[key] = str(value)
+        self._send_query_param_msg()
```

Comment:
I think we do want to support setting an int or other non-string value and have it be serialized to str. We also talked about later adding some ability to do smart typing or binding query params to non-str values in session_state - for example I have some st.number_input I want to keep synced to a give query param (Arnaud also mentioned this idea [here](https://github.com/streamlit/streamlit/issues/7665#issuecomment-1794428552) for example)

I'm not sure about the tradeoff on hinting to the developer that they will get a string back today when they do set() and then get() vs it becoming a warning. Is it ugly or more clear if I need to do `st.query_params["mykey"] = str(my_int)` to get rid of warnings? 🤔  Will it be annoying if we change the type from `str` to `Any` later due to widening the support?

Those are my thoughts - ultimately I'm OK with either decision and defer to William + Vincent on it. If I need to be the one to decide, I'd probably leave it as `Any`. But it's not strongly held.

Reply:
I think we'd still want to support setting any value and having it be serialized to `str`. The change would only be to type hints and not any actual behavior.

In this case, my hunch is that people using `mypy` with` streamlit` would prefer to just deal with these types explicitly when setting query params than have us coerce the types for them , and users that aren't using mypy might be surprised by the auto-casting, but I think that tradeoff is fine to stay consistent with how most libraries treat query params.

I think relaxing from `str` to `Any` later shouldn't cause too many issues since we'll just be allowing the setter method to accept more inputs, and the getter will still only return `str` (I'll need to think about this a bit more to be sure, though)

Reply:
Works for me